hydrolib/hydromt_delft3dfm/workflows/boundaries.py

```diff
@@ -0,0 +1,272 @@
+            dn: {**d, **{"where": "downstream"}}
+            for up, dn, d in G.edges(data=True)
+            if G.out_degree[dn] == 0 and G.degree[dn] == 1
+        }
+    elif where == "upstream":
+        endnodes = {
+            up: {**d, **{"where": "upstream"}}
+            for up, dn, d in G.edges(data=True)
+            if G.in_degree[up] == 0 and G.degree[up] == 1
+        }
+    elif where == "both":
+        endnodes = {
+            dn: {**d, **{"where": "downstream"}}
+            for up, dn, d in G.edges(data=True)
+            if G.out_degree[dn] == 0 and G.degree[dn] == 1
+        }
+        endnodes.update(
+            {
+                up: {**d, **{"where": "upstream"}}
+                for up, dn, d in G.edges(data=True)
+                if G.in_degree[up] == 0 and G.degree[up] == 1
+            }
+        )
+    else:
+        pass
+
+    if len(endnodes) == 0:
+        logger.error(f"cannot generate boundaries for given condition {where}")
+
+    endnodes_pd = (
+        pd.DataFrame().from_dict(endnodes, orient="index").drop(columns=["geometry"])
+    )
+    endnodes_gpd = gpd.GeoDataFrame(
+        data=endnodes_pd,
+        geometry=[Point(endnode) for endnode in endnodes],
+        crs=branches.crs,
+    )
+    endnodes_gpd.reset_index(inplace=True)
+    return endnodes_gpd
+
+
+def select_boundary_type(
+    boundaries: gpd.GeoDataFrame,
+    branch_type: str,
+    boundary_type: str,
+    boundary_locs: str,
+    logger=logger,
+):
+    """Select boundary location per branch type and boundary type"""
+    boundaries_branch_type = boundaries.loc[boundaries["branchType"] == branch_type, :]
+    if branch_type == "river":
+        if boundary_type == "waterlevel":
+            if boundary_locs != "both":
+                boundaries_branch_type = boundaries_branch_type.loc[
+                    boundaries_branch_type["where"] == boundary_locs, :
+                ]
+        if boundary_type == "discharge":
+            boundaries_branch_type = boundaries_branch_type.loc[
+                boundaries_branch_type["where"] == "upstream", :
+            ]
+        else:
+            logger.error(
+                f"Wrong boundary type {boundary_type} selected for {branch_type} boundaries."
+            )
+    # TODO: extend
+    # for now only downstream boundaries not connected to openwater
+    # later add connected to weir (upstream or downstream or intermediate)
+    elif branch_type == "pipe":
+        if boundary_type == "waterlevel":
+            boundaries_branch_type = boundaries_branch_type.loc[
+                boundaries_branch_type["where"] == "downstream", :
+            ]
+        else:
+            logger.error(
+                f"Wrong boundary type {boundary_type} selected for {branch_type} boundaries."
+            )
+
+    return boundaries_branch_type
+
+
+def validate_boundaries(boundaries: gpd.GeoDataFrame, branch_type: str = "river"):
+    """Validate boundaries per branch type"""
+
```

Comment:
From a software extensibility/maintainability perspective: it might be better to validate for the correct situation instead of a wrong situation.
 
For example, if we ever add/support different types of boundaries besides water level and discharge there is a big chance that we have to update this method (but will probably forget to do so, leading to possible bugs)